typescript-rewrite/samlang-core/src/interpreter/expression-interpreter.ts

```diff
@@ -0,0 +1,274 @@
+          }
+          case '==': {
+            const v1 = this.eval(expression.e1);
+            const v2 = this.eval(expression.e2);
+            if (v1.type === 'functionValue' || v2.type === 'functionValue') {
+              throw new PanicException('Cannot compare functions!');
+            }
+            return { type: 'bool', value: v1 === v2 };
+          }
+          case '!=': {
+            const v1 = this.eval(expression.e1);
+            const v2 = this.eval(expression.e2);
+            if (v1.type === 'functionValue' || v2.type === 'functionValue') {
+              throw new PanicException('Cannot compare functions!');
+            }
+            return { type: 'bool', value: v1 !== v2 };
+          }
+          case '&&': {
+            const v1 = this.eval(expression.e1) as BoolValue;
+            return !v1.value ? { type: 'bool', value: false } : this.eval(expression.e2, context);
+          }
+          case '||': {
+            const v1 = this.eval(expression.e1) as BoolValue;
+            return v1.value ? { type: 'bool', value: true } : this.eval(expression.e2, context);
+          }
+          case '::': {
+            const v1 = this.eval(expression.e1) as StringValue;
+            const v2 = this.eval(expression.e2) as StringValue;
+            return { type: 'string', value: v1.value + v2.value };
+          }
+        }
+        break;
+      case 'IfElseExpression':
+        return this.eval(
+          (this.eval(expression.e1) as BoolValue).value ? expression.e1 : expression.e2,
+          context
+        );
+      case 'MatchExpression': {
+        const matchedValue = this.eval(expression.matchedExpression, context) as VariantValue;
+        const matchedPattern =
+          expression.matchingList.find((el) => el.tag === matchedValue.tag) ??
+          this.blameTypeChecker();
+        let ctx = context;
+        if (matchedPattern.dataVariable) {
+          ctx = { ...context };
+          ctx.localValues.set(matchedPattern.dataVariable, matchedValue.data);
+        }
+        return this.eval(matchedPattern.expression, ctx);
+      }
+      case 'LambdaExpression':
+        return {
+          type: 'functionValue',
+          arguments: expression.parameters.map((param) => param[0]),
+          body: expression.body,
+          context,
+        };
+      case 'StatementBlockExpression': {
+        const { block } = expression;
+        const currentContext = context;
+        block.statements.forEach((statement) => {
+          const assignedValue = this.eval(statement.assignedExpression, currentContext);
+          const p = statement.pattern;
+          switch (p.type) {
+            case 'TuplePattern': {
+              const { tupleContent } = assignedValue as TupleValue;
+              p.destructedNames.forEach((nameWithRange, i) => {
+                if (nameWithRange[0] !== null) {
+                  currentContext.localValues.set(nameWithRange[0], tupleContent[i]);
```

Comment:
This is dangerous. Consider this example:

```kotlin
val _ = {
  val a = 5;
};
val b = a + 5;
```

This program should fail to run because `a` defined in the inner block should not escape the block.

In your implementation, it will succeed. To fix this, you need to make a defensive copy of `context` into `currentContext`.

Reply:
hmm so I did a copy on line 237 but not sure if that's what you mean?

Reply:
You need a deep copy.